### PR TITLE
rgw: add support for SSL encrypted AMQP connections

### PR DESCRIPTION
**Pull request by tschoonj**
## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_amqp.cc
Comment (yuvalif) [inline src/rgw/rgw_amqp.cc:900]: don't think we need an explicit "use-ssl" argument here. better to just follow the schema of the url and decide based on "amqp" vs "amqps"

File: src/rgw/rgw_pubsub_push.cc
Comment (yuvalif) [inline src/rgw/rgw_pubsub_push.cc:193]: see: https://github.com/ceph/ceph/pull/39392/files#r573597276
I think that we don't need this function

File: src/test/rgw/amqp_mock.cc
Comment (yuvalif) [inline src/test/rgw/amqp_mock.cc:125]: maybe add another boolean flag to "amqp_socket_t_" indicating whether this is an ssl or regular socket.
later on in the test we can validate that to make sure that the new functionality works ok

File: src/test/rgw/test_rgw_amqp.cc
Comment (yuvalif) [inline src/test/rgw/test_rgw_amqp.cc:64]: see: https://github.com/ceph/ceph/pull/39392/files#r573599891
we can probably add a simple test that verify the ssl socket was created based on the URL

**Comment by yuvalif** [comment]
in `connection_id_t` ctor: https://github.com/ceph/ceph/pull/39392/files#diff-c9ff6a7fbd82da6994645d01e3bd69d746015fac4674840f4f7b8494d861d8eaL56

we should set the default port numbers: 5672, 5671 (if not explicitly set) based on whether we use SSL or not.
otherwise, we won't be able to open 2 connections (one SSL and one not) to the same broker

**Comment by tschoonj** [comment]
Thanks for the comments, I will try and refactor accordingly tomorrow. One thing that wasn't clear to me: should I keep the `verify-ssl` option around?

**Comment by yuvalif** [comment]
> Thanks for the comments, I will try and refactor accordingly tomorrow. One thing that wasn't clear to me: should I keep the `verify-ssl` option around?

IMO, better to remove it. otherwise we have to handle contradicting conf (e.g. `verify-ssl=false` but URL schema is `amqps`)

**Comment by tschoonj** [comment]
I don't think they contradict: it is quite possible that one has to use an AMQPS only broker with a self-signed certificate and not have the corresponding CA root certificate. If using a non-SSL broker, `verify-ssl=false`, would simply be ignored. The Kafka push-endpoint also supports this, from which I copied the code.

I am also wondering if it would be useful to add an option that allows for providing the CA certificate as a string with the contents of the PEM file. Currently one would need to be able to install the CA certificate on all the Ceph nodes, before it can be used. At work I cannot do this, as I am just a Ceph user, and not an admin.

**Comment by yuvalif** [comment]
> I don't think they contradict: it is quite possible that one has to use an AMQPS only broker with a self-signed certificate and not have the corresponding CA root certificate. If using a non-SSL broker, `verify-ssl=false`, would simply be ignored. The Kafka push-endpoint also supports this, from which I copied the code.
> 
> I am also wondering if it would be useful to add an option that allows for providing the CA certificate as a string with the contents of the PEM file. Currently one would need to be able to install the CA certificate on all the Ceph nodes, before it can be used. At work I cannot do this, as I am just a Ceph user, and not an admin.

sorry, my bad :-( 
you are absolutely right. i mean that we can remove the `use-ssl` flag, and rely on the schema.
please keep the `verify-ssl` flag, it is very useful, especially for test environments.

**Comment by tschoonj** [comment]
Question: how do I test only the `rgw_amqp` unit tests? `make check` takes almost an hour to complete...

**Comment by yuvalif** [comment]
> Question: how do I test only the `rgw_amqp` unit tests? `make check` takes almost an hour to complete...

you can use `ctest -R amqp` to run tests based on regular expression. or just execute the specific binary: `bin/unittest_rgw_amqp`

**Comment by tschoonj** [comment]
Thanks for that!

I amended the commit per your comments.

**Comment by tschoonj** [comment]
I am now looking into adding a proper unittest to `tests_ps.py`.

I followed the [instructions](https://docs.ceph.com/en/latest/dev/developer_guide/running-tests-locally/#running-test-using-vstart-runner-py) to run individual python unittests but got quickly into trouble.

After setting `skip_push_tests` to `False`, I ran (after running `vstart.sh`):

```bash
python ../qa/tasks/vstart_runner.py tasks.rgw_multi.tests_ps.test_ps_s3_notification_push_amqp_on_master
Using guessed paths /home/awf63395/github/ceph/build/lib/ ['/home/awf63395/github/ceph/qa', '/home/awf63395/github/ceph/build/lib/cython_modules/lib.3', '/home/awf63395/github/ceph/src/pybind']
Using guessed paths /home/awf63395/github/ceph/build/lib/ ['/home/awf63395/github/ceph/qa', '/home/awf63395/github/ceph/build/lib/cython_modules/lib.3', '/home/awf63395/github/ceph/src/pybind']

              RabbitMQ 3.6.10. Copyright (C) 2007-2017 Pivotal Software, Inc.
  ##  ##      Licensed under the MPL.  See http://www.rabbitmq.com/
  ##  ##
  ##########  Logs: /var/log/rabbitmq/rabbit@rfi-advanced-chemistry-mnova-vm.log
  ######  ##        /var/log/rabbitmq/rabbit@rfi-advanced-chemistry-mnova-vm-sasl.log
  ##########
              Starting broker...
 completed with 0 plugins.
Traceback (most recent call last):
  File "../qa/tasks/vstart_runner.py", line 1768, in <module>
    exec_test()
  File "../qa/tasks/vstart_runner.py", line 1584, in exec_test
    max_required_mgr, require_memstore = scan_tests(modules)
  File "../qa/tasks/vstart_runner.py", line 1303, in scan_tests
    overall_suite = load_tests(modules, loader.TestLoader())
  File "../qa/tasks/vstart_runner.py", line 1292, in load_tests
    module_suites.append(loader.loadTestsFromName(mod_name))
  File "/usr/lib/python3.6/unittest/loader.py", line 204, in loadTestsFromName
    test = obj()
  File "/home/awf63395/github/ceph/qa/tasks/rgw_multi/tests_ps.py", line 1398, in test_ps_s3_notification_push_amqp_on_master
    master_zone, _ = init_env(require_ps=False)
  File "/home/awf63395/github/ceph/qa/tasks/rgw_multi/tests_ps.py", line 578, in init_env
    zonegroup = realm.master_zonegroup()
AttributeError: 'NoneType' object has no attribute 'master_zonegroup'
```

Any thoughts?

Thanks in advance!

**Comment by yuvalif** [comment]
> I am now looking into adding a proper unittest to `tests_ps.py`.
> 
> I followed the [instructions](https://docs.ceph.com/en/latest/dev/developer_guide/running-tests-locally/#running-test-using-vstart-runner-py) to run individual python unittests but got quickly into trouble.
> 
> After setting `skip_push_tests` to `False`, I ran (after running `vstart.sh`):
> 
> ```shell
> python ../qa/tasks/vstart_runner.py tasks.rgw_multi.tests_ps.test_ps_s3_notification_push_amqp_on_master
> Using guessed paths /home/awf63395/github/ceph/build/lib/ ['/home/awf63395/github/ceph/qa', '/home/awf63395/github/ceph/build/lib/cython_modules/lib.3', '/home/awf63395/github/ceph/src/pybind']
> Using guessed paths /home/awf63395/github/ceph/build/lib/ ['/home/awf63395/github/ceph/qa', '/home/awf63395/github/ceph/build/lib/cython_modules/lib.3', '/home/awf63395/github/ceph/src/pybind']
> 
>               RabbitMQ 3.6.10. Copyright (C) 2007-2017 Pivotal Software, Inc.
>   ##  ##      Licensed under the MPL.  See http://www.rabbitmq.com/
>   ##  ##
>   ##########  Logs: /var/log/rabbitmq/rabbit@rfi-advanced-chemistry-mnova-vm.log
>   ######  ##        /var/log/rabbitmq/rabbit@rfi-advanced-chemistry-mnova-vm-sasl.log
>   ##########
>               Starting broker...
>  completed with 0 plugins.
> Traceback (most recent call last):
>   File "../qa/tasks/vstart_runner.py", line 1768, in <module>
>     exec_test()
>   File "../qa/tasks/vstart_runner.py", line 1584, in exec_test
>     max_required_mgr, require_memstore = scan_tests(modules)
>   File "../qa/tasks/vstart_runner.py", line 1303, in scan_tests
>     overall_suite = load_tests(modules, loader.TestLoader())
>   File "../qa/tasks/vstart_runner.py", line 1292, in load_tests
>     module_suites.append(loader.loadTestsFromName(mod_name))
>   File "/usr/lib/python3.6/unittest/loader.py", line 204, in loadTestsFromName
>     test = obj()
>   File "/home/awf63395/github/ceph/qa/tasks/rgw_multi/tests_ps.py", line 1398, in test_ps_s3_notification_push_amqp_on_master
>     master_zone, _ = init_env(require_ps=False)
>   File "/home/awf63395/github/ceph/qa/tasks/rgw_multi/tests_ps.py", line 578, in init_env
>     zonegroup = realm.master_zonegroup()
> AttributeError: 'NoneType' object has no attribute 'master_zonegroup'
> ```
> 
> Any thoughts?
> 
> Thanks in advance!

these tests are part of the multisite suite, so they start the multi zone cluster themselves.
this means that you should not run `vstart.sh` (use `src/stop.sh` to stop the vstart cluster you started).
under the hood, these tests are running `mstart.sh` but this does not need to be called explicitly.

* first you would need to have a file holding the cluster setup. e.g. `test_multi.conf`
```
[DEFAULT]
num_zonegroups=1
num_zones=1
num_ps_zones=0
gateways_per_zone=1
no_bootstrap=false
log_level=5
```
* assuming you put this file under you `build` directory, and run from there, you can execute the tests by:
```
MON=1 OSD=1 MGR=0 MSD=0 RGW_MULTI_TEST_CONF=./test_multi.conf nosetests -s ../src/test/rgw/test_multi.py:<name of your test>
```
* if you want to run the test again, you would need to skip the setup phase of the test, so, in `test_multi.conf` you would need to change:
```
no_bootstrap=true
```
note that you don't need to restart the cluster if you change the python test code.
* if you want to restart the cluster (e.g. after code change in the RGW), you should first stop it, by calling `../src.mstop.sh c1` and then run the test again (with `no_bootstrap=false` in the conf file).

official instructions
 are [here](https://github.com/ceph/ceph/blob/master/src/test/rgw/test_multi.md) but probably missing some stuff.

**Review by yuvalif — APPROVED** [review]

**Comment by tschoonj** [comment]
Wait, I haven't tested any of this with a real RabbitMQ server!

Also, I need to update the docs

**Comment by tschoonj** [comment]
To test against an external RabbitMQ server with LetsEncrypt certificates, I guess I need to contact the Ceph endpoint over https as I am sending a username and password across as well to create the topic.

So I tried using:

```python
    topic_conf = PSTopicS3(master_zone.secure_conn, topic_name, zonegroup.name, endpoint_args=endpoint_args)
```

instead of:

```python
    topic_conf = PSTopicS3(master_zone.conn, topic_name, zonegroup.name, endpoint_args=endpoint_args)
```

But then I run into:

```
Traceback (most recent call last):
  File "/home/awf63395/github/ceph/build/venv/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/awf63395/github/ceph/build/venv/lib/python3.6/site-packages/nose/util.py", line 620, in newfunc
    return func(*arg, **kw)
  File "/home/awf63395/github/ceph/src/test/rgw/rgw_multi/tests_ps.py", line 3559, in test_ps_s3_creation_triggers_on_master_external
    external_endpoint_address=os.environ['AMQP_EXTERNAL_ENDPOINT'])
  File "/home/awf63395/github/ceph/src/test/rgw/rgw_multi/tests_ps.py", line 3503, in ps_s3_creation_triggers_on_master
    topic_conf = PSTopicS3(master_zone.secure_conn, topic_name, zonegroup.name, endpoint_args=endpoint_args)
  File "/home/awf63395/github/ceph/src/test/rgw/rgw_multi/zone_ps.py", line 203, in __init__
    protocol = 'https' if conn.is_secure else 'http'
AttributeError: 'NoneType' object has no attribute 'is_secure'
```

Any advice on how I can ensure that the endpoint can be reached over https?

Thanks!

**Comment by yuvalif** [comment]
> To test against an external RabbitMQ server with LetsEncrypt certificates, I guess I need to contact the Ceph endpoint over https as I am sending a username and password across as well to create the topic.
> 
> So I tried using:
> 
> ```python
>     topic_conf = PSTopicS3(master_zone.secure_conn, topic_name, zonegroup.name, endpoint_args=endpoint_args)
> ```
> 
> instead of:
> 
> ```python
>     topic_conf = PSTopicS3(master_zone.conn, topic_name, zonegroup.name, endpoint_args=endpoint_args)
> ```
> 
> But then I run into:
> 
> ```
> Traceback (most recent call last):
>   File "/home/awf63395/github/ceph/build/venv/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
>     self.test(*self.arg)
>   File "/home/awf63395/github/ceph/build/venv/lib/python3.6/site-packages/nose/util.py", line 620, in newfunc
>     return func(*arg, **kw)
>   File "/home/awf63395/github/ceph/src/test/rgw/rgw_multi/tests_ps.py", line 3559, in test_ps_s3_creation_triggers_on_master_external
>     external_endpoint_address=os.environ['AMQP_EXTERNAL_ENDPOINT'])
>   File "/home/awf63395/github/ceph/src/test/rgw/rgw_multi/tests_ps.py", line 3503, in ps_s3_creation_triggers_on_master
>     topic_conf = PSTopicS3(master_zone.secure_conn, topic_name, zonegroup.name, endpoint_args=endpoint_args)
>   File "/home/awf63395/github/ceph/src/test/rgw/rgw_multi/zone_ps.py", line 203, in __init__
>     protocol = 'https' if conn.is_secure else 'http'
> AttributeError: 'NoneType' object has no attribute 'is_secure'
> ```
> 
> Any advice on how I can ensure that the endpoint can be reached over https?
> 
> Thanks!

you need to restart the cluster with `use_ssl=true` in your `test_multi.conf` file.

...we really should fix the documentation around that :-(

**Comment by tschoonj** [comment]
So, I got a bit further after your latest comment regarding turning on SSL on the ceph endpoint.

The changes I push right now reflect my attempts at getting it to work with my Letsencrypt SSL certificates enabled RabbitMQ server.

So it basically works but it's problematic.

1. rabbitmq-c currently does not load the CA certificates that may be installed on the system. There is an easy fix for that in the rabbitmq-c source code, which I [suggested](https://github.com/alanxz/rabbitmq-c/issues/648#issuecomment-778155297) to the developer. This means that my Letsencrypt certificate results in an error as soon as the connection is established as they cannot be verified.
2. I could also get Ceph to load those certificates itself, by fetching the SSL context from the RabbitMQ SSL socket with `amqp_ssl_socket_get_context`, available in the latest release 0.10.0, but which has not seen widespread adoption among the major Linux releases. For example it was not backported to Ubuntu Bionic (which I am using for development at the moment), and is nowhere to be seen on even the latest releases of RHEL and CentOS.
3. It should still be possible to load the certificates in Ceph using a rather ugly hack, based on this [piece of code](https://github.com/alanxz/rabbitmq-c/pull/560), which drove its author to contribute `amqp_ssl_socket_get_context` to rabbitmq-c. By checking the rabbitmq-c version macros at compile time, one could easily use `amqp_ssl_socket_get_context`  when rabbitmq-c >= 0.10.0, and the ugly hack otherwise. 
4. I eventually managed to get the events sent to my RabbitMQ server over SSL, but with verification disabled. The test that I added for this passes nicely.

Next, I will look into writing a test with a local rabbitmq server that uses self-signed SSL certificates...

**Comment by yuvalif** [comment]
> So, I got a bit further after your latest comment regarding turning on SSL on the ceph endpoint.
> 
> The changes I push right now reflect my attempts at getting it to work with my Letsencrypt SSL certificates enabled RabbitMQ server.
> 
> So it basically works but it's problematic.
> 
> 1. rabbitmq-c currently does not load the CA certificates that may be installed on the system. There is an easy fix for that in the rabbitmq-c source code, which I [suggested](https://github.com/alanxz/rabbitmq-c/issues/648#issuecomment-778155297) to the developer. This means that my Letsencrypt certificate results in an error as soon as the connection is established as they cannot be verified.
> 2. I could also get Ceph to load those certificates itself, by fetching the SSL context from the RabbitMQ SSL socket with `amqp_ssl_socket_get_context`, available in the latest release 0.10.0, but which has not seen widespread adoption among the major Linux releases. For example it was not backported to Ubuntu Bionic (which I am using for development at the moment), and is nowhere to be seen on even the latest releases of RHEL and CentOS.
> 3. It should still be possible to load the certificates in Ceph using a rather ugly hack, based on this [piece of code](https://github.com/alanxz/rabbitmq-c/pull/560), which drove its author to contribute `amqp_ssl_socket_get_context` to rabbitmq-c. By checking the rabbitmq-c version macros at compile time, one could easily use `amqp_ssl_socket_get_context`  when rabbitmq-c >= 0.10.0, and the ugly hack otherwise.
> 4. I eventually managed to get the events sent to my RabbitMQ server over SSL, but with verification disabled. The test that I added for this passes nicely.
> 
> Next, I will look into writing a test with a local rabbitmq server that uses self-signed SSL certificates...

would be great if you do (3), bu the basic requirement would be that it work with a self-signed certificate.
note sure why it would be limited to local machine? can't you have certs on the different machines use the same root cert?

**Comment by tschoonj** [comment]
Cool will look into implementing that next week then.

It should already work with self-signed certificates, but to confirm that I need to write the unittest with an appropriate local rabbitmq instance.

**Comment by tschoonj** [comment]
This was easier than I thought 😄 

I can now connect to my RabbitMQ server with SSL verification

**Comment by tschoonj** [comment]
I have written a test using self-signed certificates. It works fine, but comes currently with the limitation that ceph needs to be restarted before it can be run again.

This is due to the connection remaining registered within Ceph even after the RabbitMQ server is killed. At this point the log starts filling up with:

```
2021-02-13T20:08:15.584+0000 7f3599ffb700 20 AMQP run: retry connection
2021-02-13T20:08:15.584+0000 7f3599ffb700 10 AMQP run: connection (localhost:5671/) retry failed. error: RGW_AMQP_STATUS_SOCKET_OPEN_FAILED (-9)
```

When running the test again, which comes with a new SSL certificate, things go wrong, because the certificate tied to the previous connection is not replaced with the new one.

How long do connections to a RabbitMQ server that no longer can be reached continue to exist? IMHO they should be deleted after a short time.

**Comment by yuvalif** [comment]
> I have written a test using self-signed certificates. It works fine, but comes currently with the limitation that ceph needs to be restarted before it can be run again.
> 
> This is due to the connection remaining registered within Ceph even after the RabbitMQ server is killed. At this point the log starts filling up with:
> 
> ```
> 2021-02-13T20:08:15.584+0000 7f3599ffb700 20 AMQP run: retry connection
> 2021-02-13T20:08:15.584+0000 7f3599ffb700 10 AMQP run: connection (localhost:5671/) retry failed. error: RGW_AMQP_STATUS_SOCKET_OPEN_FAILED (-9)
> ```
> 
> When running the test again, which comes with a new SSL certificate, things go wrong, because the certificate tied to the previous connection is not replaced with the new one.
> 
> How long do connections to a RabbitMQ server that no longer can be reached continue to exist? IMHO they should be deleted after a short time.

currently we retry the same connection indefinitely...
first step in fix in fixing that would be to destroy connection when no more topics need it: https://tracker.ceph.com/issues/49033
this would allow re-creation of the connection via re-creation of the topic.
automatic re-creation of the connection (after some timeout) is also possible, but, may be trickier to implement correctly

**Comment by tschoonj** [comment]
> currently we retry the same connection indefinitely...
first step in fix in fixing that would be to destroy connection when no more topics need it: https://tracker.ceph.com/issues/49033
this would allow re-creation of the connection via re-creation of the topic.
automatic re-creation of the connection (after some timeout) is also possible, but, may be trickier to implement correctly

Well I think that should go into a separate PR. 

Is there anything else, aside from documentation updates, that you need me to do before this PR can be merged in?

**Comment by yuvalif** [comment]
> > currently we retry the same connection indefinitely...
> > first step in fix in fixing that would be to destroy connection when no more topics need it: https://tracker.ceph.com/issues/49033
> > this would allow re-creation of the connection via re-creation of the topic.
> > automatic re-creation of the connection (after some timeout) is also possible, but, may be trickier to implement correctly
> 
> Well I think that should go into a separate PR.
> 

agree

> Is there anything else, aside from documentation updates, that you need me to do before this PR can be merged in?

from your end, only documentation.
the PR will also have to pass regression (this is beside the CI unit tests check)

**Comment by tschoonj** [comment]
Do you know why that first make check test fails? It doesn't seem to be related to my changes in this PR.

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
> Do you know why that first make check test fails? It doesn't seem to be related to my changes in this PR.

unrelated. kicked another run

**Comment by tschoonj** [comment]
I updated the docs now.

By the way to get my tests working I had to apply the following patch as well:

```diff
diff --git a/src/mrun b/src/mrun
index 3122342e1b..23302f2503 100755
--- a/src/mrun
+++ b/src/mrun
@@ -21,10 +21,10 @@ fi
 
 shift 2
 
-if [ "$RGW_VALGRIND" == 'yes' ] && [ $command == 'radosgw' ]; then
-    valgrind --trace-children=yes --tool=memcheck --max-threads=1024 $CEPH_BIN/$command -c $CEPH_CONF_PATH/ceph.conf "$@"
-    sleep 10
-else
+#if [ "$RGW_VALGRIND" == 'yes' ] && [ $command == 'radosgw' ]; then
+#    valgrind --trace-children=yes --tool=memcheck --max-threads=1024 $CEPH_BIN/$command -c $CEPH_CONF_PATH/ceph.conf "$@"
+#    sleep 10
+#else
     $CEPH_BIN/$command -c $CEPH_CONF_PATH/ceph.conf "$@"
     $CEPH_BIN/$command -c $CEPH_CONF_PATH/ceph.conf "$@"
-fi
+#fi
 
diff --git a/src/test/rgw/rgw_multi/tests_ps.py b/src/test/rgw/rgw_multi/tests_ps.py
index 2a82f2c0ad..4ebb1fad5b 100644
--- a/src/test/rgw/rgw_multi/tests_ps.py
+++ b/src/test/rgw/rgw_multi/tests_ps.py
@@ -38,7 +38,7 @@ import boto.s3.tagging
 # configure logging for the tests module
 log = logging.getLogger(__name__)
 
-skip_push_tests = True
+skip_push_tests = False
 
 ####################################
 # utility functions for pubsub tests
@@ -613,6 +613,7 @@ def get_ip():
     """ This method returns the "primary" IP on the local box (the one with a default route)
     source: https://stackoverflow.com/a/28950776/711085
     this is needed because on the teuthology machines: socket.getfqdn()/socket.gethostname() return 127.0.0.1 """
+    return "localhost"
     s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
     try:
```

Ubuntu's bash didn't like those brackets for some reason. Also I had to make `get_ip` return `localhost`, as the returned IP address was rejected by RabbitMQ when using it with the default passwords.

**Comment by yuvalif** [comment]
> Ubuntu's bash didn't like those brackets for some reason. 

feel free to submit a tracker issue for that

> Also I had to make `get_ip` return `localhost`, as the returned IP address was rejected by RabbitMQ when using it with the default passwords.

don't see any issue with using "localhost", but you can workarounbd the issue by setting:
```
loopback_users=none
```
in the rabbitmq conf file
see: https://www.rabbitmq.com/access-control.html#loopback-users

File: src/rgw/rgw_amqp.cc
Comment (yuvalif) [inline src/rgw/rgw_amqp.cc:410]: according to this: https://github.com/alanxz/rabbitmq-c/blob/master/librabbitmq/amqp_ssl_socket.h#L68
it is supported since v0.9.0

File: src/test/rgw/amqp_mock.cc
Comment (yuvalif) [inline src/test/rgw/amqp_mock.cc:148]: see: https://github.com/ceph/ceph/pull/39392/files#r575828688

File: src/test/rgw/amqp_mock.cc
Comment (yuvalif) [inline src/test/rgw/amqp_mock.cc:149]:
```suggestion
void* amqp_ssl_socket_get_context(amqp_socket_t *self) {
```
gcc10 complains about ambiguity

File: src/test/rgw/rgw_multi/tests_ps.py
Comment (yuvalif) [inline src/test/rgw/rgw_multi/tests_ps.py:3516]: we should support the non-secure case as well:
```python
PSTopicS3(master_zone.conn, topic_name, zonegroup.name, endpoint_args=endpoint_args)
```

File: src/test/rgw/rgw_multi/tests_ps.py
Comment (yuvalif) [inline src/test/rgw/rgw_multi/tests_ps.py:3607]: can we use a hardcoded root key, so we won't need to restart the RGW on every run?

File: src/rgw/rgw_amqp.cc
Comment (tschoonj) [reply src/rgw/rgw_amqp.cc:410]: It actually is 0.10.0, the comment there is wrong.

Check the [Changelog](https://github.com/alanxz/rabbitmq-c/releases/tag/v0.10.0). The [commit](https://github.com/alanxz/rabbitmq-c/commit/60adf5f8817f32b25a99aa54ba186a29208c7942) responsible for adding that method is marked as being present in 0.10.0 and master only.

File: src/test/rgw/amqp_mock.cc
Comment (tschoonj) [reply src/test/rgw/amqp_mock.cc]: Fixed

File: src/test/rgw/rgw_multi/tests_ps.py
Comment (tschoonj) [reply src/test/rgw/rgw_multi/tests_ps.py]: Fixed. Secure connections will now only be used with connections to external AMQP servers.

File: src/test/rgw/rgw_multi/tests_ps.py
Comment (tschoonj) [reply src/test/rgw/rgw_multi/tests_ps.py:3610]: I tried that, even with all keys and certificates hardcoded, but still didn't work.

File: src/test/rgw/rgw_multi/tests_ps.py
Comment (yuvalif) [inline src/test/rgw/rgw_multi/tests_ps.py:3516]:
```suggestion
    if ca_location:
```
thought that `ca_location` is the indication for secure connection?

Comment (tschoonj) [reply src/test/rgw/rgw_multi/tests_ps.py:3516]: A secure connection to the ceph endpoint is only required when passing a rabbitmq username and password, which you will almost certainly will need for an external RabbitMQ instance.

For a RabbitMQ running on localhost with guest/guest authentication you don't need a secure connection with Ceph, even if you're going to connect to RabbitMQ over SSL.

Comment (yuvalif) [reply src/test/rgw/rgw_multi/tests_ps.py:3516]: RGW should reject the topic creation if it sees user/password sent in cleartext

Comment (tschoonj) [reply src/test/rgw/rgw_multi/tests_ps.py:3516]: It does, but in the test with the self-signed certificates to connect to the RabbitMQ server on localhost, I am not sending a username/password across, as the guest account will be used.

File: src/test/rgw/rgw_multi/tests_ps.py
Comment (yuvalif) [reply src/test/rgw/rgw_multi/tests_ps.py:3610]: ok. i guess we will have to look at it again once we automate these tests

**Comment by yuvalif** [comment]
jenkins test api

**Comment by tschoonj** [comment]
Ready for a merge?

**Comment by yuvalif** [comment]
> Ready for a merge?

waiting for all runs to finish: http://pulpito.front.sepia.ceph.com/yuvalif-2021-02-15_17:36:22-rgw:multisite-wip-yuval-fix-42902-distro-basic-gibba/

**Comment by yuvalif** [comment]
> > Ready for a merge?
> 
> waiting for all runs to finish: http://pulpito.front.sepia.ceph.com/yuvalif-2021-02-15_17:36:22-rgw:multisite-wip-yuval-fix-42902-distro-basic-gibba/

we can probably ignore the 2 "queued" tests. could be considered a "pass"

@cbodley can you please merge if no other issues?

**Comment by cbodley** [comment]
@tschoonj thanks for the contribution!

@yuvalif should we plan on backports? feel free to update https://tracker.ceph.com/issues/42902

**Comment by yuvalif** [comment]
> @tschoonj thanks for the contribution!
> 
> @yuvalif should we plan on backports? feel free to update https://tracker.ceph.com/issues/42902

added "pacific". @tschoonj do you need that in octopus as well?

**Comment by tschoonj** [comment]
> added "pacific". @tschoonj do you need that in octopus as well?

Would be great if you could do that!

**Pull request merged**
